### ISSUE TITLE
Segfault with nightly in python 3.9

### ISSUE THREAD

**Issue by ppwwyyxx** [pytorch/pytorch#50014]
## 🐛 Bug

A simple code segfaults in nightly builds with python 3.9 but works with python 3.8
## To Reproduce
Content of `Dockerfile`:
```
FROM python:3.9-slim-buster
# works with 3.8
USER root
WORKDIR /root
RUN apt-get update && apt-get install git build-essential --yes
RUN python3 -m pip install --upgrade pip
RUN python3 -m pip install --pre torch==1.8.0.dev20210102+cpu  -f https://download.pytorch.org/whl/nightly/cpu/torch_nightly.html
RUN python3 -m pip install Pillow
RUN git clone https://github.com/pytorch/vision/
ENV PYTHONPATH=./vision
COPY test-backbone.py /
RUN python /test-backbone.py
```

Content of `test-backbone.py`:
```
from torchvision.models import resnet50
import torch


if __name__ == "__main__":
    backbone = resnet50(pretrained=True)

    img = torch.randn((2, 3, 256, 288))
    features = backbone(img)
    loss = features.sum()
    loss.backward()
    print("exiting")
```

Run `docker build  -t py39:v0 .`

Logs:
```
Step 11/11 : RUN python /test-backbone.py
 ---> Running in 31a88ed05b3d
Downloading: "https://download.pytorch.org/models/resnet50-19c8e357.pth" to /root/.cache/torch/hub/checkpoints/resnet50-19c8e357.pth
100.0%
exiting
Segmentation fault (core dumped)
The command '/bin/sh -c python /test-backbone.py' returned a non-zero code: 139

```

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser

**Comment by malfet** [comment]
I have the repro, sounds very similar to the original Python-3.9 issue with PyBind:
* thread #18, name = 'python', stop reason = signal SIGSEGV: invalid address (fault address: 0x555600000483)
  * frame #0: 0x0000555555754d7d python`PyThreadState_Clear at pystate.c:785
    frame #1: 0x00007fffe94f98ec libtorch_python.so`pybind11::gil_scoped_acquire::dec_ref() + 44
    frame #2: 0x00007fffe94f9929 libtorch_python.so`pybind11::gil_scoped_acquire::~gil_scoped_acquire() + 9

**Comment by SamuelMarks** [comment]
FYI: With the current release version of PyTorch I'm getting a segfault after my tests run, the issue itself appears to be finalizing threads at exit time, it looks like it's related to autograd, I guess that's where the offending threads are coming from, so likely a problem with pybind11 related to the GIL.

https://github.com/SamuelMarks/ml-params-pytorch/runs/1698382105

[Nothing new, this is more of a MeToo post]

**Comment by mattip** [comment]
Duplicate of others, the base issue seems to be #38228

**Comment by dstein64** [comment]
I reported a similar issue a few hours ago in #50858, which I'll close now as a duplicate.

Information I reported in that ticket, which may be relevant here as well, is 1) that I only encountered the problem on Ubuntu, but not on Windows or macOS, as shown [here](https://github.com/dstein64/pytorch-50858/actions/runs/500472568), and 2) the problem occurs on `v1.7.1`, the current release (in addition to the nightly build mentioned above in this ticket).

Here's the code I was using to trigger the problem, which was generated as a minimal reproducible example (i.e., not the actual code where I first encountered the segmentation fault on exit).

```python
import torch
import torchvision.models as models

print('start')
resnet18 = models.resnet18(pretrained=True)
loss = torch.tensor([1.0], requires_grad=True)
loss.backward()
print('end')
```

**Comment by SamuelMarks** [comment]
@dstein64 Interesting. I'll experiment with my Ubuntu box later, but possibly a `del resnet18` would also do the trick…

**Comment by dstein64** [comment]
> "but possibly a del resnet18 would also do the trick…"

@SamuelMarks, in the example code above, `del`ing `resnet18` and/or `loss` still results in a segmentation fault.

**Comment by LoryPack** [comment]
I too have similar problem when running unittests for a Python package on travis, using Python 3.9. Specifically, all unit tests run fine, but the final outcome is segfault. See: https://travis-ci.org/github/LoryPack/abcpy/builds/755321073

The problem there occurred with both versions 1.7.1 and 1.7.0
